player/javascript.c

```diff
@@ -0,0 +1,1237 @@
+// This can throw, but if it does, then it's after the cleanup.
+// Macro because the callee could have any prototype.
+#define AW(expr) do {\
+    autowash *aw = new_autowash();\
+    if (js_try(J)) {\
+        delete_autowash(aw);\
+        js_throw(J);\
+    }\
+    expr;\
+    js_endtry(J);\
+    delete_autowash(aw);\
+} while (0)
+
+// Prototype for autowash functions which can be called from inside the vm.
+typedef void (*aw_CFunction)(js_State*, autowash*);
+
+// The trampoline function through which all autowash functions are called from
+// inside the vm. Obtains the target function address and autowash-call it.
+static void script__autowash(js_State *J)
+{
+    // The target function is at the "aw_" property of this function instance.
+    js_currentfunction(J);
+    js_getproperty(J, -1, "aw_");
+    aw_CFunction fn = (aw_CFunction)js_touserdata(J, -1, "aw_fn");
+    js_pop(J, 2);
+
+    AW(fn(J, aw));
+}
+
+// Identical to js_newcfunction, but the function is inserted with an autowash
+// wrapper, and its prototype should have the additional aw argument.
+static void aw_newcfunction(js_State *J, aw_CFunction fn, const char *name,
+                            int length)
+{
+    js_newcfunction(J, script__autowash, name, length);
+    js_pushnull(J);  // a prototype for the userdata object
+    js_newuserdata(J, "aw_fn", fn, NULL);  // uses a "aw_fn" verification tag
+    js_defproperty(J, -2, "aw_", JS_READONLY | JS_DONTENUM | JS_DONTCONF);
+}
+
+/**********************************************************************
+ *  Initialization and file loading
+ *********************************************************************/
+
+static const char *get_builtin_file(const char *name)
+{
+    for (int n = 0; builtin_files[n][0]; n++) {
+        if (strcmp(builtin_files[n][0], name) == 0)
+            return builtin_files[n][1];
+    }
+    return NULL;
+}
+
+// Pushes the content of fname: from builtin_files, else from the OS.
+static void aw_push_file(js_State *J, const char *fname, autowash *aw)
+{
+    char *filename = mp_get_user_path(aw, jctx(J)->mpctx->global, fname);
+    MP_VERBOSE(jctx(J), "Reading file '%s'\n", filename);
+
+    const char *builtin = get_builtin_file(filename);
+    if (builtin) {
+        js_pushstring(J, builtin);
+        return;
+    }
+
+    FILE *f = aw->f = fopen(filename, "rb");
+    if (!f)
+        js_error(J, "cannot open file: '%s'", filename);
+
+    if (fseek(f, 0, SEEK_END) < 0)
+        js_error(J, "cannot seek in file: '%s'", filename);
+
+    long n = ftell(f);
+    if (n < 0)
+        js_error(J, "cannot read file size: '%s'", filename);
+
+    if (fseek(f, 0, SEEK_SET))
+        js_error(J, "cannot seek in file: '%s'", filename);
+
+    char *s = talloc_array(aw, char, n);
+    if (!s)
+        js_error(J, "cannot allocate %ld bytes for file: '%s'", n, filename);
+
+    int t = fread(s, 1, n, f);
+    if (t != n)
+        js_error(J, "cannot read data from file: '%s'", filename);
```

Comment:
I'll change it to read till `EOF`, and (quoting from IRC) "<wm4> if it blows up, caller's fault".